### PR TITLE
gobjectIntrospection: fix incorrect GIR shared library paths

### LINKED ISSUE #34988: meson sometimes generates incorrect GIR paths
For some packages, a wrong path to the library is generated. For example, `fwupd` produces an extra `libfwupd` directory:

```ShellSession
$ cat $(nix-build -A fwupd)/share/gir-1.0/Fwupd-2.0.gir | grep libfwupd
             shared-library="/nix/store/mnjbmk19k7r4256d7gqdlf6iqbi089y7-fwupd-1.0.4/lib/libfwupd/libfwupd.so.2"
```

Now, the typelib path is supposed to be addressed by https://github.com/NixOS/nixpkgs/pull/31453. It only works when typelibdir is an absolute path but apparently that is not the problem here (it is [already absolute](https://github.com/hughsie/fwupd/blob/6a029ba2f6d13f5bc9433aeb947da35a5f07e6b1/meson.build#L236)). Possibly a bug in the [gobject-introspecton patch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch)?

This is how the gir file is built (add `ninjaFlags = "-v";` to `fwupd` expression):

```
[39/223] /nix/store/275wyhj6wf72dlx0806jxnpn0sxl43nv-gobject-introspection-1.54.1-dev/bin/g-ir-scanner -I/nix/store/275wyhj6wf72dlx0806jxnpn0sxl43nv-gobject-introspection-1.54.1-dev/include/gobject-introspection-1.0 -I/nix/store/yz0kpbxsb3lxzy4mz4mj50s9sm4456gn-glib-2.54.3-dev/include/glib-2.0 -I/nix/store/9bjbfcwb73ayga2bm3p0crzfa1nfzmb7-glib-2.54.3/lib/glib-2.0/include --no-libtool --namespace=Fwupd --nsversion=2.0 --warn-all --output libfwupd/Fwupd-2.0.gir -I/build/fwupd-1.0.4/libfwupd -I/build/fwupd-1.0.4/build/libfwupd -I./libfwupd/.. -I../libfwupd/.. --filelist=/build/fwupd-1.0.4/build/libfwupd/fwupd@sha/Fwupd_2.0_gir_filelist --include=Gio-2.0 --include=GObject-2.0 --include=Soup-2.4 --symbol-prefix=fwupd --identifier-prefix=Fwupd --pkg-export=fwupd --cflags-begin -fstack-protector-strong -Waggregate-return -Wunused -Warray-bounds -Wcast-align -Wclobbered -Wdeclaration-after-statement -Wduplicated-cond -Wempty-body -Wformat=2 -Wformat-nonliteral -Wformat-security -Wformat-signedness -Wignored-qualifiers -Wimplicit-function-declaration -Winit-self -Wlogical-op -Wmissing-declarations -Wmissing-format-attribute -Wmissing-include-dirs -Wmissing-noreturn -Wmissing-parameter-type -Wmissing-prototypes -Wnested-externs -Wno-error=cpp -Wno-discarded-qualifiers -Wno-missing-field-initializers -Wno-strict-aliasing -Wno-suggest-attribute=format -Wno-unused-parameter -Wnull-dereference -Wold-style-definition -Woverride-init -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wstrict-aliasing -Wstrict-prototypes -Wswitch-default -Wtype-limits -Wundef -Wuninitialized -Wunused-but-set-variable -Wunused-variable -Wwrite-strings -D_DEFAULT_SOURCE -DFWUPD_DISABLE_DEPRECATED -I./libfwupd/.. -I../libfwupd/.. -I/nix/store/yz0kpbxsb3lxzy4mz4mj50s9sm4456gn-glib-2.54.3-dev/include/gio-unix-2.0/ -I/nix/store/yz0kpbxsb3lxzy4mz4mj50s9sm4456gn-glib-2.54.3-dev/include/glib-2.0 -I/nix/store/9bjbfcwb73ayga2bm3p0crzfa1nfzmb7-glib-2.54.3/lib/glib-2.0/include -I/nix/store/smw2d05fjna0nq930fv4ppgqlh7lmv48-libsoup-2.60.2-dev/include/libsoup-2.4 -I/nix/store/gz5alm6sfd87xn3k82y9xg74xzx37jby-json-glib-1.4.2-dev/include/json-glib-1.0 --cflags-end -L/build/fwupd-1.0.4/build/libfwupd -L/nix/store/jarc6ksyzqkwa4qpxs5dg07cfqzagw8m-json-glib-1.4.2/lib --extra-library=gobject-2.0 --extra-library=gio-2.0 -L/nix/store/qag0629fzp901skyyvrjjyxh4ga66j62-libsoup-2.60.2/lib --extra-library=soup-2.4 -L/nix/store/9bjbfcwb73ayga2bm3p0crzfa1nfzmb7-glib-2.54.3/lib --extra-library=json-glib-1.0 --extra-library=glib-2.0 --library fwupd
g-ir-scanner: link: gcc -o /build/fwupd-1.0.4/build/tmp-introspectY0mIpA/Fwupd-2.0 /build/fwupd-1.0.4/build/tmp-introspectY0mIpA/Fwupd-2.0.o -L. -Wl,-rpath,. -Wl,--no-as-needed -lfwupd -lgobject-2.0 -lgio-2.0 -lsoup-2.4 -ljson-glib-1.0 -lglib-2.0 -L/build/fwupd-1.0.4/build/libfwupd -Wl,-rpath,/build/fwupd-1.0.4/build/libfwupd -L/nix/store/jarc6ksyzqkwa4qpxs5dg07cfqzagw8m-json-glib-1.4.2/lib -Wl,-rpath,/nix/store/jarc6ksyzqkwa4qpxs5dg07cfqzagw8m-json-glib-1.4.2/lib -L/nix/store/qag0629fzp901skyyvrjjyxh4ga66j62-libsoup-2.60.2/lib -Wl,-rpath,/nix/store/qag0629fzp901skyyvrjjyxh4ga66j62-libsoup-2.60.2/lib -L/nix/store/9bjbfcwb73ayga2bm3p0crzfa1nfzmb7-glib-2.54.3/lib -Wl,-rpath,/nix/store/9bjbfcwb73ayga2bm3p0crzfa1nfzmb7-glib-2.54.3/lib -L/nix/store/9bjbfcwb73ayga2bm3p0crzfa1nfzmb7-glib-2.54.3/lib -lgio-2.0 -lgobject-2.0 -Wl,--export-dynamic -lgmodule-2.0 -pthread -lglib-2.0
```

## Technical details

Nixpkgs 0d20bf0287c349ff889d71c772468470026038cb

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
When a library path contained the library name it was eagerly matched

    libfwupd.so.2 => /build/fwupd-1.0.5/build/libfwupd/libfwupd.so.2 (0x00007ffff7bbd000)
                                              ^^^^^^^^^^^^^^^^^^^^^^
    libgweather-3.so.15 => /build/libgweather-3.28.0/build/libgweather/libgweather-3.so.15 (0x00007ffff7bae000)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

which lead to a broken shared library path in the generated GIR file.

This patch allows the soname on the left-hand side of the arrow to
be matched to avoid the trap of the right-hand side. A negative
lookahead had to be added to select the store path, since only
the first match is taken into account.

    libglib-2.0.so.0 => /nix/store/dqlc8y4phlg1hmdbwkhqfwhnxcac88d1-glib-2.56.0/lib/libglib-2.0.so.0 (0x00007ffff6400000)

This will not fix non-GNU platforms, where the soname is not printed
first, but we cannot do much without structured ldd output.

Closes: https://github.com/NixOS/nixpkgs/issues/34988

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @hedning

### CODE REVIEW

**Comment by jtojnar** [comment]
Cherry picked for easy reviewing, does not apply to gobject introspection in master though.

File: pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch [hunk 1]
```diff
@@ -53,57 +51,50 @@ index 89ec193..54f1d2e 100755
...
++                    m = pattern.search(line)
                  if m:
                      del patterns[library]
 -                    shlibs.append(m.group(1))
-+                    match = m.group(1)
-+                    if not match.startswith('/') \
-+                       and len(options.fallback_libpath) > 0:
-+                        match = os.path.join(options.fallback_libpath, match)
-+                    shlibs.append(match)
++                    shlibs.append(os.path.join(options.fallback_libpath, m.group(1)))
```
Comment (hedning) [inline pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch]: I'm guessing `fallback_libpath` is known to be empty when dealing with absolute paths?

Comment (jtojnar) [reply pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch]: > If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component.

https://docs.python.org/3/library/os.path.html#os.path.join

Comment (hedning) [reply pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch]: Ah, that makes sense.

**Review by hedning — APPROVED** [review]
LGTM, tested the _ldd_nix_pattern function minimally, and it seemed to do the job :)

**Pull request merged**